cluster-api/pkg/controller/machineset/controller.go

```diff
@@ -0,0 +1,218 @@
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Info("shutting down machineset controller")
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false
+// when it's time to quit.
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncHandler holds the controller's business logic.
+// it makes sure that the current state is equal to the desired state.
+// note that the current state of the cluster is calculated based on the number of machines
+// that are owned by the given machineSet (key).
+func (c *Controller) syncHandler(key string) error {
+	machineSet, err := c.machineSetsLister.Get(key)
+	if err != nil {
+		return err
+	}
+
+	filteredMachines, err := c.getMachines(machineSet)
+	if err != nil {
+		return err
+	}
+
+	return c.syncReplicas(machineSet, filteredMachines)
+}
+
+// syncReplicas essentially scales machine resources up and down.
+func (c *Controller) syncReplicas(machineSet *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
+	currentMachineCount := int32(len(machines))
+	desiredReplicas := int32(0)
+	if machineSet.Spec.Replicas != nil {
+	 desiredReplicas = *machineSet.Spec.Replicas
+	}
+
+	if desiredReplicas > currentMachineCount {
+		glog.V(2).Infof("creating a machine ( spec.replicas(%d) > currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
+		machine, err := c.createMachine(machineSet)
+		if err != nil {
+			return err
+		}
+		_, err = c.machineClient.ClusterV1alpha1().Machines().Create(machine)
+		if err != nil {
+			return err
+		}
+		if err = waitForMachineInCache(machine.Name, c.machineLister); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if desiredReplicas < currentMachineCount {
+		glog.V(2).Infof("deleting a machine ( spec.replicas(%d) < currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
+		machineToDelete := machines[0].Name
+		err := c.machineClient.ClusterV1alpha1().Machines().Delete(machineToDelete, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		if err = waitForMachineDeletedInCache(machineToDelete, c.machineLister); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return nil
+}
+
+// createMachine creates a machine resource.
+// the name of the newly created resource is created as machineSet.Name + some random characters.
+// in addition, the newly created resource is owned by the given machineSet (we set the OwnerReferences field)
+func (c *Controller) createMachine(machineSet *clusterv1alpha1.MachineSet) (*clusterv1alpha1.Machine, error) {
+	machineName := machineSet.Name + "-" + string(uuid.NewUUID())[:6]
```

Comment:
At some point we might be prepared for a name conflict here and handle it.

Reply:
POSTing a machine with the same name will yield `HTTP` 409 error, hopefully on the next run we will generate a different name. If my memory serves me well we can generate 32 to the power of 6 different names. Which gives us a decent number :)

Reply:
We could use the `metadata.generateName` field so the apiserver takes care of generating a name.

Reply:
Good suggestion, @mrIncompetent. @p0lyn0mial do you think you can incorporate this suggestion?

Reply:
yeah, good idea.